src/ccache.c

```diff
@@ -0,0 +1,523 @@
+}
+
+/* Flushes half of the bin */
+static bool
+ccache_bin_flush(tsd_t *tsd, ccache_t *ccache, ccache_bin_t *bin,
+    arena_t *arena, szind_t binind, void *ptr, bool small) {
+	const unsigned nflush = CCACHE_BIN_ELEMENTS / 2;
+	assert(ccache_bin_locked(bin));
+
+	atomic_fetch_add_u64(&ccache->stats.nflushes, 1,
+	    atomic_memory_order_relaxed);
+
+	CACHE_BIN_PTR_ARRAY_DECLARE(ptrs, nflush);
+	/* 
+	 * TODO: tcache is flushed differently - 'oldest' ptrs are flushed and
+	 * the rest is memmoved to the beginning of the bin. It helps first fit,
+	 * apparently. Here I pick the easiest way - flush the 'top' half
+	 * (leftmost in the address space). Other policies should be tested as
+	 * well.
+	 */
+	ptrs.ptr = bin->ccache_bin_entry;
+
+	arena_flush(tsd, arena, &ptrs, binind, nflush, NULL, small,
+	    /* ccache = */ true);
+
+	bin->ccache_bin_entry[nflush - 1] = ptr;
+	__asm__ __volatile__ ("" ::: "memory" );
+	bin->head = &bin->ccache_bin_entry[nflush - 1];
+
+	/* TODO: re-consider return true or false on success. This caused a bug
+	 * becuase the return value was forwarded by ccache_free, which should
+	 * return true on success, but this was returning 'false'
+	 */
+	return true;
+}
+
+static void
+ccache_bin_full_flush_unsafe(tsd_t *tsd, ccache_bin_t *bin, szind_t binind,
+    bool small) {
+	const unsigned nflush = ccache_bin_ncached_elements(bin);
+
+	CACHE_BIN_PTR_ARRAY_DECLARE(ptrs, nflush);
+	ptrs.ptr = &bin->ccache_bin_entry[CCACHE_BIN_ELEMENTS - nflush];
+
+	arena_flush(tsd, *tsd_arenap_get(tsd), &ptrs, binind, nflush, NULL,
+	    small, true);
+
+	bin->head = (void *)&bin->head;
+}
+
+static void
+ccache_bump_tstats(tsd_t *tsd, szind_t binind) {
+	cache_bin_stats_t *stats = ccache_bin_tstats_get(tsd, binind);
+	++stats->nrequests;
+}
+
+/* Depends on booted tcache */
+bool
+ccache_init(tsdn_t *tsdn, base_t *base) {
+	assert(ncpus > 0);
+
+	ccache_minind = sz_size2index(tcache_maxclass) + 1;
+	ccache_maxind = ccache_minind + ccached_nclasses - 1;
+	ccache_maxclass = sz_index2size(ccache_maxind);
+	ccache_bin_size = PAGE;
+	ccache_size = sizeof(ccache_t) * ncpus;
+	ccache = (ccache_t*)base_alloc(tsdn, base, ccache_size,
+	    PAGE);
+	if (ccache == NULL) {
+		return true;
+	}
+	for (unsigned i = 0; i < ncpus; ++i) {
+		ccache_t *ith_cache = &ccache[i];
+		for (unsigned j = 0; j < ccached_nclasses; ++j) {
+			ccache_bin_t *bin = &ith_cache->bins[j];
+			bin->head = (void **)&bin->head;
+		}
+	}
+
+	return false;
+}
+
+void*
+ccache_alloc(tsd_t *tsd, arena_t *arena, size_t size, szind_t ind,
+    bool zero, bool small) {
+	/*
+	 * Ccache is disabled for manual arenas due to difficulties handling
+	 * arena reset.
+	 */
+	assert(arena_is_auto(arena));
+
+	void *ret = NULL;
+	volatile rseq_t *rseq_abi = &tsd_ccache_tdatap_get(tsd)->rseq_abi;
+	int cpu;
+	ccache_t *ccache;
+	ccache_bin_t *bin;
+restart:
+	cpu = rseq_abi->cpu_id;
+	ccache = ccache_get(cpu);
+	bin = ccache_bin_get(ccache, ind);
+	__asm__ __volatile__ goto(
```

Comment:
What would happen if this runs on a host w/o rseq supported kernel?

Reply:
There's TODO in `tsd_ccache_init` - the initialization should fail. I still haven't implemented the handler for this case. I currently picture it as follows: either we disable percpu cache in runtime, or abort if `opt.abort` is `true`.